src/utils/socksProxy.ts

```diff
@@ -0,0 +1,329 @@
+    // +----+-----+-------+------+----------+----------+
+    // |VER | CMD |  RSV  | ATYP | DST.ADDR | DST.PORT |
+    // +----+-----+-------+------+----------+----------+
+    // | 1  |  1  | X'00' |  1   | Variable |    2     |
+    // +----+-----+-------+------+----------+----------+
+
+    // Response.
+    // +----+-----+-------+------+----------+----------+
+    // |VER | REP |  RSV  | ATYP | BND.ADDR | BND.PORT |
+    // +----+-----+-------+------+----------+----------+
+    // | 1  |  1  | X'00' |  1   | Variable |    2     |
+    // +----+-----+-------+------+----------+----------+
+
+    const version = await this._readByte();
+    assert(version === 0x05, 'The VER field must be set to x05 for this version of the protocol, was ' + version);
+
+    const command = await this._readByte();
+    await this._readByte();  // skip reserved.
+    const addressType = await this._readByte();
+    let host = '';
+    switch (addressType) {
+      case SocksAddressType.IPv4:
+        host = (await this._readBytes(4)).join('.');
+        break;
+      case SocksAddressType.FqName:
+        const length = await this._readByte();
+        host = (await this._readBytes(length)).toString();
+        break;
+      case SocksAddressType.IPv6:
+        const bytes = await this._readBytes(16);
+        const tokens = [];
+        for (let i = 0; i < 8; ++i)
+          tokens.push(bytes.readUInt16BE(i * 2));
+        host = tokens.join(':');
+        break;
+    }
+    const port = (await this._readBytes(2)).readUInt16BE(0);
+
+    this._buffer = Buffer.from([]);
+    this._offset = 0;
+    this._fence = 0;
+
+    return {
+      command,
+      host,
+      port
+    };
+  }
+
+  private async _readByte(): Promise<number> {
+    const buffer = await this._readBytes(1);
+    return buffer[0];
+  }
+
+  private async _readBytes(length: number): Promise<Buffer> {
+    this._fence = this._offset + length;
+    if (!this._buffer || this._buffer.length < this._fence)
+      await new Promise<void>(f => this._fenceCallback = f);
+    this._offset += length;
+    return this._buffer.slice(this._offset - length, this._offset);
+  }
+
+  private _writeBytes(buffer: Buffer) {
+    if (this._socket.writable)
+      this._socket.write(buffer);
+  }
+
+  private _onClose() {
+    this._client.onSocketClosed(this._uid);
+  }
+
+  private _onData(buffer: Buffer) {
+    this._buffer = Buffer.concat([this._buffer, buffer]);
+    if (this._fenceCallback && this._buffer.length >= this._fence) {
+      const callback = this._fenceCallback;
+      this._fenceCallback = undefined;
+      callback();
+    }
+  }
+
+  setLocalAddress(host: string, port: number) {
+    this._writeBytes(Buffer.from([
+      0x05,
+      SocksReply.Succeeded,
+      0x00, // RSV
+      0x01, // IPv4
+      ...parseIP(host), // Address
+      port << 8, port & 0xFF // Port
+    ]));
+    this._socket.on('data', data => this._client.onSocketData(this._uid, data));
+  }
+
+  sendError(code: string) {
```

Comment:
be careful with non-successful replies in WebKit on Linux.

Reply:
Passes tests now.